WordPress/gutenberg
# Focus control in widgets customizer

## Description

Close #30897.

Handle focus control in the Widgets Customizer. We listen to the `focus-control-for-setting` event globally and handle expanding the section and focusing on the block.

I also revamped the structure a little bit, instead of mounting and unmounting a block editor React tree for each sidebar, we mount one React tree globally (_virtually_) and use `createPortal` to handle rendering the actual editor to each sidebar. We still have to unmount and remount the editor when switching to another sidebar though, the performance does not improve much. However, I hope that the code is bit more structured and flexible now. One benefit of this approach is that we can now use React context to pass down states down the tree. `<MoveToSidebar>` can now be written in a more _React-way_ (if that's a thing).

Also upgrade `puppeteer-testing-library` to v0.5.0 (full [changelog](https://github.com/kevin940726/puppeteer-testing-library/compare/v0.4.0...v0.5.0)). This enable us to wait for `toHaveFocus()` to finish.

## How has this been tested?



Added tests in `customizing-widgets`.

- Click on pencil icon should expand the corresponding widget area and focus the block.
- Click on the same pencil icon again should still focus the block.
- Click on another pencil icon in the same widget area should move focus to the other block.
- Click on another pencil icon in a different widget area should expand the widget area and move focus to the block.

## Screenshots 

https://user-images.githubusercontent.com/7753001/116508897-e6a52000-a8f4-11eb-9c40-af42596fccaf.mp4


## Types of changes




Bug fix

## Checklist:
